### ISSUE TITLE
Support or document how to use multiple credentials

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/pub#1471]
I have a personal `@mydomain.com` account and `@corp.com` account.

Right now I have a shell script that swaps between two files for `~/.pub-cache/credentials.json`.

Instead, how about:

```bash
$ pub publish --auth my@email.com
```

(I'd take a shot at helping to implement this)

**Comment by nex3** [comment]
This seems reasonable to add. It'll require a fair amount of UI design, though—we'll need a way for users to add, remove, and list the credentials they have saved.

**Comment by matanlurey** [comment]
Here is my mini-design with limited hand waving:

## Design

Introduce a new `pub auth` command, similar to `pub uploader`.

```bash
$ pub help auth
Manages authorization credentials for editing a package on pub.dartlang.org.

Usage: pub auth {add/remove/list/clear} <email>
```

### `pub auth add`

```bash
$ pub auth add

Authorize by clicking on this URL: https://<the-current-pub-authorization-url>
```

Upon success credentials are stored in `<pub-cache-root>/credentials.json`:

```json
{
  "matan-personal@gmail.com": {
    "accessToken": "...",
    "tokenEndpoint": "...",
    "scopes": "...",
    "expiration": "...",
  },
}
```

### `pub auth remove`

```bash
$ pub auth remove matan-corp@corpemail.com

Deleted authorization credentials for "matan-corp@corpemail.com". You will no longer be able to upload or edit packages as "matan-corp@corpemail.com" until you authorize again using `pub auth add`.
```

The key in `credentials.json` is deleted at this point.

### `pub auth list`

```bash
$ pub auth list

You are authenticated as: 
- matan-personal@gmail.com
- <abc-order-of-all-email-accounts>
```

### `pub auth clear`

```bash
$ pub auth clear

Are you sure you want to remove all credentials for editing and uploading packages on pub.dartlang.org? You'll need to re-add each credential using `pub auth add`.

> Are you sure [Y/N]?
```

If `Y`, we could just delete or clear out `credentials.json`.

## Release

Once we introduce this feature, it could be conditional, i.e., until you try to add more than one account we use the existing code and format of the `credentials.json`. We could also automatically "upgrade" them from the flat file format to the "many-credentials-supported" format when they run `pub auth` for the first time.

The first time a user without credentials types `pub lish`, it would be similar to the current flow, they will be prompted to login. The difference is the credentials will be added in the nested format, not the flat format.

**Comment by nex3** [comment]
Which set of credentials are the default?

**Comment by matanlurey** [comment]
@nex3 Given that authorized users is public, could we:

1. If there is only a single matching authorized user, use that one.
2. If there is multiple matching authorized users (probably very rare), use most recent*

*I'd be tempted to add a way to set defaults, but I can't imagine this is common enough - and most users are going to be concerned with it just working - not which user actually does the uploading - right?

**Comment by nex3** [comment]
I think having some notion of a default is important—I don't want to make people explicitly choose their account each time they publish a new package. Given that, it may not be worth the extra effort of trying to sniff the correct account as opposed to just trying to use the default and allowing the user to specify an alternative if they want.

**Comment by matanlurey** [comment]
@nex3: Is this a default _per_ package, or just a global default?

If this is a global default, then we do need a flag for `pub lish` and `pub uploader`, i.e.:

```bash
$ pub lish --use_credentials matan-corp@corpemail.com
```

Otherwise users will have to switch their defaults constantly for one-off packages.

Otherwise SGTM.

**Comment by nex3** [comment]
> Is this a default _per_ package, or just a global default?

A global default. Implicit per-package configuration sounds like a huge can of worms :stuck_out_tongue_closed_eyes:.

> If this is a global default, then we do need a flag for `pub lish` and `pub uploader`, i.e.:
> 
> ```bash
> $ pub lish --use_credentials matan-corp@corpemail.com
> ```
> 
> Otherwise users will have to switch their defaults constantly for one-off packages.

:+1:

Sounds like we've got a design! I'm going to un-assign myself since I don't have the cycles to implement it right now, but if you (or anyone else) wants to I'm happy to review.

**Comment by matanlurey** [comment]
> > Is this a default _per_ package, or just a global default?
> 
> A global default. Implicit per-package configuration sounds like a huge can of worms :stuck_out_tongue_closed_eyes:.

Whew.

> 
> > If this is a global default, then we do need a flag for `pub lish` and `pub uploader`, i.e.:
> > 
> > ```bash
> > $ pub lish --use_credentials matan-corp@corpemail.com
> > ```
> > 
> > Otherwise users will have to switch their defaults constantly for one-off packages.
> 
> :+1:
> 
> Sounds like we've got a design! I'm going to un-assign myself since I don't have the cycles to implement it right now, but if you (or anyone else) wants to I'm happy to review.

Sgtm. I plan on working this in the background between other tasks - would you be still available to review the PRs?

**Comment by nex3** [comment]
I'm always happy to review :smiley:.

**Comment by matanlurey** [comment]
Given that I think @mit-mit's team might help with future features here, I'm going to wait and see what he/his team thinks before starting on any new feature work. Still interested though!

**Comment by leonsenft** [comment]
I just stumbled into this; thanks for mentioning the `~/.pub-cache/credientials.json` trick! I was authenticated with the wrong credentials, and couldn't find any official documentation on how to change them.

**Comment by gamebox** [comment]
I also recently ran into issues with this.  Has there been any movement here?  @matanlurey ?  I'd love to help with design or implementation if this is something that we'd like to have.

**Comment by jonasfj** [comment]
@sigurdm recently landed `pub logout` in https://github.com/dart-lang/pub/pull/2035, I'm not sure how much more we should support.

Isn't it simple to log-out and log-in again? I imagine that most people will work around this by:
 * using different computers for work and personal accounts, or,
 * add the work account as uploader on personal packages (having both personal and work account as uploader on such packages).

Do we need this?

**Comment by zoechi** [comment]
Perhaps an optional `--cred-name=foo` parameter that allows to login and logout and publish as a specific user where `cred-name` is used for naming `~/.pub-cache/foo-credientials.json`

**Comment by filiph** [comment]
FWIW, I just added a self-answered question on StackOverflow ([here](https://stackoverflow.com/questions/55671876/how-do-i-log-out-from-darts-pub-package-manager/55671877#55671877)) that tries to document this. The version of `pub` with the `logout` command hasn't reached me yet (I'm on Dart stable, v2.2.0).

**Comment by sigurdm** [comment]
This is a very old issue.

I think we won't do this. The advice is gonna be:

* Either manually switch between credential files or
* Use `dart pub logout` and login with the right user or
* [Publish from github](https://dart.dev/tools/pub/automated-publishing), and avoid having the credentials locally at all :)
